---
title: "Final Code for Paper"
output: html_document
date: '2022-07-15'
---

This notebook generates the following figures:

1. Fig. S6A
2. Fig. S6B

This notebook requires the following inputs to run:

1. simulated influenza .csv files
2. flu.synthetic.afdata.csv (.csv containing variant information)
3. maf_functions.R (set of functions written by Katherine Johnson to process .vcf data)

Load libraries
```{r}
library("tidyr")
library("plyr")
library('ggplot2')
library('dplyr')
library("plotrix") 
library("plotly")
library("glue")
library('vcfR')
library('tools')
library('seqinr')

source(glue("/Users/marissaknoll/Desktop/MAF/Paper/maf_functions.R"))
```

# Simulated Influenza Data

Set working directory and file names
```{r}
wkdir = "/Users/marissaknoll/Desktop/MAF/Paper"
fig_dir = glue("{wkdir}/Figures")

h1n1_af_report = "Reports/H1N1-7_agitated_thompson_af_data.csv" # h1n1
h3n2_af_report = "Reports/H3N2-4_thirsty_volta_af_data.csv" # h3n2
vict_af_report = "Reports/Vic-3_fervent_raman_af_data.csv" # vic
sars_af_report = "Reports/SARS-CoV2-3_kickass_mcclintock_af_data.csv" # sars

#segment_sizes = "Metadata/SegmentSize.maf.csv"

cov_cut = 300
freq_cut = 0.02
ntlist = c("A","C","T","G")
COVERAGE = c("100","200","300","500","1000","10000","100000")
```

Read in simulated data
```{r}
h1n1_af = read.csv(h1n1_af_report,header = T)
h3n2_af = read.csv(h3n2_af_report,header = T)
vict_af = read.csv(vict_af_report,header = T)
sars_af = read.csv(sars_af_report, header = T)

#sizes = read.csv(segment_sizes,header = T)
```

Combine data from different viruses
```{r}
h1n1_af$strain = "H1N1"
h3n2_af$strain = "H3N2"
vict_af$strain = "VICT"
sars_af$strain = "SARS"

AV = rbind(h1n1_af, h3n2_af, vict_af,sars_af)
```

Rearrange data & extract information from sample_id column
```{r}
vdf = rearrange_output_data(AV)

vdf$CHROM = vdf$chrom
vdf = vdf %>% separate(CHROM, c(NA, "segment"))

vdf$segment = factor(vdf$segment, levels = c("PB2","PB1","PA","HA","NP","NA","MP","NS","CoV2"))

timo = filter(vdf, tool == "timo")
timo$parameter2[is.na(timo$parameter2)] = "bino"
timo = timo %>% filter(parameter == "standard" & parameter2 == "bino")
# only working with timo with binocheck on

freebayes = filter(vdf, tool == "freebayes")
freebayes = filter(freebayes, parameter == "custom")
#swapping freebayes custom and standard - was breaking for Allison

mutect2 = filter(vdf,tool == "mutect2")
mutect2 = filter(mutect2, parameter == "standard" & parameter2 == "unfiltered")

vdf = vdf %>% filter(tool != "timo" & tool!= "freebayes" & tool != "mutect2") %>%
  filter(parameter == "standard")
  
vdf = rbind(vdf,timo,freebayes, mutect2)
vdf$parameter2[is.na(vdf$parameter2)] = "NONE"  # have to add so other na"s can be turned to zero
  
# removing all filtered data - only working with unfiltered
# an error will pop up because we are separating a name that doesn't have all comp

vdf = select(vdf, !(c(AF,frac,align)))
vdf = filter(vdf, ref %in% ntlist & alt %in% ntlist)
```

Check number of SNPs
```{r}
locate_input_snps = vdf %>% filter(af_golden > 0 ) %>%
                        select(chrom, pos, ref, alt, strain,segment, rep) %>% unique()

locate_input_snps %>% group_by(strain, rep) %>% add_tally(name = 'NumberOfSNPs') %>% ungroup() %>%
  unique() %>% select(strain, NumberOfSNPs) %>% unique()
```

Separate replicates
```{r}
vdf_m1 = filter(vdf, rep == "m1") %>% droplevels()
vdf_m2 = filter(vdf, rep == "m2") %>% droplevels()
vdf_mx = filter(vdf, rep == "mx") %>% droplevels()
```

Add TP, FN, FP info
```{r}
vdf_m1 = add_category_information(vdf_m1)
vdf_m1$coverage = as.numeric(as.character(vdf_m1$coverage))

vdf_m2 = add_category_information(vdf_m2)
vdf_m2$coverage = as.numeric(as.character(vdf_m2$coverage))

vdf_mx = add_category_information(vdf_mx)
vdf_mx$coverage = as.numeric(as.character(vdf_mx$coverage))
```

Select true posiitves and calucate percent error
```{r}
tp = rbind(vdf_m1, vdf_mx)
tp = filter(tp, category == "TP") %>% 
  droplevels()

tp$perc_err = abs((tp$af_workflow - tp$af_golden) / tp$af_golden) * 100
```

Plot change in accuracy of AF with two reps - using random AF data only
```{r}
tp$tool.rep = paste0(tp$tool,"_",tp$rep)
tp_sub = filter(tp, coverage == "100" | coverage == "500" | coverage == "1000")

Fig_S6A = ggplot(tp_sub, aes(x = tool.rep, y = perc_err, color = tool)) +
  #geom_point(alpha = 0.3) +
  geom_boxplot(position = "dodge", outlier.shape = NA, aes(group = tool.rep, fill = rep)) +
  facet_grid(~coverage) +
  ylab("Percent Error") +
  ylim(0,100) +
  PlotTheme1 +
  caller_colScale +
  scale_fill_manual(values = c("#e6e6e6","#ffffff"))
print(Fig_S6A)
ggsave("Fig_S6A_mk.png",Fig_S6A,width = 10, height = 5)
ggsave("Fig_S6A_mk.pdf",Fig_S6A,width = 10, height = 5, useDingbats = FALSE)
```

# Synthetic Influenza Data

Set new working directory
```{r}
wkdir_syn = "/Users/marissaknoll/Desktop/MAF/Paper/SynData"
setwd(setwd(wkdir)) 
```

Load data and filter out poor quality samples
```{r}
FluSyn = glue("{wkdir_syn}/flu.synthetic.afdata.csv")

sample_remove= c('0.0039_10^3_rep2',
                  '0.0078_10^3_rep2',
                  '0.0156_10^3_rep2',
                  '0.0313_10^3_rep2',
                  '0.0625_10^7_rep1',
                  '0.25_10^7_rep1',
                  '0.5_10^7_rep1')

syn_vars = read.csv(FluSyn) %>% 
  filter(copy_number != "10^7" & !sample %in% sample_remove) %>% 
  droplevels() %>% unique()

syn_vars = syn_vars[!duplicated(syn_vars), ] %>% droplevels()
syn_vars$ID = paste0(syn_vars$expected_freq,"_",syn_vars$copy_number)

syn_tp = filter(syn_vars, cat == "TP")
```

Caluclate median frequency (rather than expected frequency)
```{r}
syn_tp_median = syn_tp %>% 
  group_by(expected_freq, copy_number,segment) %>% 
  summarise(medianfreq = median(varfreq))

syn_tp_wmed = merge(syn_tp, syn_tp_median, by = c("expected_freq", "copy_number","segment"))
syn_tp_wmed = syn_tp_wmed %>% filter(!copy_number == "control")
```

Separate replicates and find intersection
```{r}
syn_rep1 = filter(syn_tp_wmed, Rep == "rep1") %>% droplevels()
syn_rep2 = filter(syn_tp_wmed, Rep == "rep2") %>% droplevels()

syn_reps = merge(syn_rep1, syn_rep2, 
                 by = c("segment","ntpos","expected_freq","tool","copy_number","refnt","varnt",
                        "golden_REF","golden_ALT","cat","ID","medianfreq"))
syn_reps = syn_reps[!duplicated(syn_reps), ] %>% droplevels()
```

Orgainze data to combine single rep and two rep information
```{r}
syn_reps$varfreq = (syn_reps$varfreq.x + syn_reps$varfreq.y) /2
syn_reps$Rep = "repM"

sub_syn_reps = select(syn_reps, c("segment","ntpos","expected_freq","copy_number","tool","refnt", 
                                  "varnt","golden_REF","golden_ALT","cat","medianfreq", "varfreq","Rep"))
sub_syn_rep1 = select(syn_rep1, c("segment","ntpos","expected_freq","copy_number","tool","refnt", 
                                  "varnt","golden_REF","golden_ALT","cat","medianfreq", "varfreq","Rep"))

comp_syn = rbind(sub_syn_rep1, sub_syn_reps)
```

Caluclate percent error
```{r}
comp_syn$varfreq = as.numeric(comp_syn$varfreq)
comp_syn$medianfreq = as.numeric(comp_syn$medianfreq)

comp_syn$perc_err = (abs(comp_syn$varfreq - comp_syn$medianfreq) / comp_syn$medianfreq) * 100

comp_syn$tool.rep = paste0(comp_syn$tool,"_",comp_syn$Rep)
comp_syn$Rep = factor(comp_syn$Rep, levels = c("rep1","repM"))
```

Make figures
```{r}
Fig_S6B = ggplot(comp_syn, aes(x = tool.rep, y = perc_err, color = tool)) +
  #geom_point(alpha = 0.3) +
  geom_boxplot(position = "dodge", outlier.shape = NA, aes(group = tool.rep, fill = Rep)) +
  facet_grid(copy_number~expected_freq) +
  ylab("Percent Error") +
  ylim(0,100) +
  PlotTheme1 +
  caller_colScale +
  rep_colScale_fill
print(Fig_S6B)
ggsave("Fig_S6B.pdf", Fig_S6B, path = fig_dir, useDingbats = FALSE, width = 20, height = 20)

Fig_S6B_5 = ggplot(filter(comp_syn, copy_number == "10^5"), aes(x = tool.rep, y = perc_err, color = tool)) +
  #geom_point(alpha = 0.3) +
  geom_boxplot(position = "dodge", outlier.shape = NA, aes(group = tool.rep, fill = Rep)) +
  facet_grid(~expected_freq) +
  ylab("Percent Error") +
  ylim(0,100) +
  PlotTheme1 +
  caller_colScale +
  rep_colScale_fill
print(Fig_S6B_5)
ggsave("Fig_S6B_only10^5.pdf", Fig_S6B_5, path = fig_dir, useDingbats = FALSE, width = 15, height = 10)
```